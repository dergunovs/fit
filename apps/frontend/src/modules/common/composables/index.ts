import { ref, computed, onMounted } from 'vue';
import { useRouter, useRoute } from 'vue-router';

import LayoutAdmin from '@/common/components/LayoutAdmin.vue';
import LayoutDefault from '@/common/components/LayoutDefault.vue';

export function useLayout() {
  const router = useRouter();
  const route = useRoute();

  const isLoaded = ref(false);

  const layoutComponent = computed(() => (route.meta.layout === 'admin' ? LayoutAdmin : LayoutDefault));

  onMounted(async () => {
    await router.isReady();
    isLoaded.value = true;
  });

  return {
    isLoaded,
    layoutComponent,
  };
}

export function useRouteId(title: string, isQuery?: boolean) {
  const route = useRoute();

  const id = computed(() => {
    const value = isQuery ? route.query[title]?.toString() : route.params[title].toString();

    return value || '';
  });

  return { id };
}
