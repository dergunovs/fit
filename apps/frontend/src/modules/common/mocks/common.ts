import { vi } from 'vitest';

export const spyCopyToClipboard = vi.spyOn(navigator.clipboard, 'writeText').mockResolvedValue();
