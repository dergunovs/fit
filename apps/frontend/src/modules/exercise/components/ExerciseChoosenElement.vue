<template>
  <div :class="$style.exercise">
    <span data-test="exercise-title">{{ exerciseTitle }}</span>

    <UiFlex>
      <UiButton v-if="props.isSetCreatable" @click="emit('createSet')" layout="plain" data-test="exercise-create-set">
        +{{ t('set.super') }}
      </UiButton>

      <UiClose
        v-if="props.exercise._id"
        @click="emit('delete', props.exercise._id)"
        isSmall
        data-test="exercise-delete"
      />
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import { DefaultLocaleMessageSchema, useI18n } from 'vue-i18n';
import { IExerciseChoosen, TLocale } from 'fitness-tracker-contracts';
import { UiButton, UiClose, UiFlex } from 'mhz-ui';
import { localeField } from 'mhz-helpers';

interface IProps {
  exercise: IExerciseChoosen;
  index: number;
  isSetCreatable: boolean;
}

interface IEmit {
  delete: [id: string];
  createSet: [];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { t, locale } = useI18n<DefaultLocaleMessageSchema, TLocale>();

const exerciseTitle = computed(
  () => `${props.index}. ${props.exercise.exercise?.[localeField('title', locale.value)]} x${props.exercise.repeats}`
);
</script>

<style module lang="scss">
.exercise {
  display: flex;
  gap: 16px;
  align-items: center;
  justify-content: space-between;
  padding: 6px 12px;
  background-color: var(--color-gray-light);
  border-radius: 16px;
}
</style>
