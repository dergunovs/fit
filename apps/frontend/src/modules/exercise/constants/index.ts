export const URL_EXERCISE = '/exercises';
export const URL_EXERCISE_CREATE = `${URL_EXERCISE}/create`;
export const URL_EXERCISE_EDIT = `${URL_EXERCISE}/edit`;

export const EXERCISE_REPEATS_OPTIONS = [8, 10, 12, 15, 20];
export const EXERCISE_REPEATS_DEFAULT = 12;

export const EXERCISE_STATISTICS_HEADERS = [{ title: 'Упражнение' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const MUSCLE_GROUP_STATISTICS_HEADERS = [{ title: 'Мышцы' }, { title: 'Подходы' }, { title: 'Повторы' }];
export const EXERCISE_LIST_HEADERS = [{ title: 'Название' }];

export const EXERCISE_DELETED_TITLE = 'Упражнение удалено';
