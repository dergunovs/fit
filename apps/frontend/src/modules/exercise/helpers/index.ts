import {
  EXERCISE_MUSCLE_GROUPS,
  IExercise,
  IExerciseChoosen,
  IExerciseDone,
  IExerciseStatistics,
  IMuscleGroup,
  IUser,
  IUserEquipment,
} from 'fitness-tracker-contracts';

import { ITimelineStep } from '@/activity/interface';
import { getWeightsForUserEquipment } from '@/equipment/helpers';

function isUserHasEquipment(exercise: IExercise, user?: IUser | null) {
  return user?.equipments?.some((equipment) => equipment.equipment?.title === exercise.equipment?.title);
}

function isUserHasEquipmentForWeight(exercise: IExercise, user?: IUser | null) {
  return user?.equipments?.some((equipment) =>
    exercise.equipmentForWeight?.some((equipmentForWeight) => equipmentForWeight.title === equipment.equipment?.title)
  );
}

function checkIsUserEquipmentMatches(exercise: IExercise, user?: IUser | null) {
  let result = false;

  const isExerciseHasEquipment = exercise.equipment;
  const isExerciseHasEquipmentForWeight = exercise.equipmentForWeight?.length;
  const isWeightsRequired = exercise.isWeightsRequired;

  if (!isExerciseHasEquipment && !isExerciseHasEquipmentForWeight) {
    result = true;
  } else if (isUserHasEquipment(exercise, user)) {
    result = true;
  } else if (
    !isExerciseHasEquipment &&
    isExerciseHasEquipmentForWeight &&
    isUserHasEquipmentForWeight(exercise, user)
  ) {
    result = true;
  } else if (!isExerciseHasEquipment && !isWeightsRequired) {
    result = true;
  }

  return result;
}

export function getAverageDuration(exercise: IExerciseStatistics, type: 'set' | 'repeat') {
  return type === 'set'
    ? `${((exercise.averageDuration * exercise.repeats) / exercise.sets || 0).toFixed(1)}с`
    : `${exercise.averageDuration.toFixed(1)}с`;
}

export function generateMuscleGroupStatistics(exercises: IExerciseDone[]) {
  const groups: {
    title: string;
    sets: number;
    repeats: number;
  }[] = [];

  EXERCISE_MUSCLE_GROUPS.forEach((group: IMuscleGroup) => {
    const title = group.title;
    let sets = 0;
    let repeats = 0;

    exercises.forEach((exercise: IExerciseDone) => {
      const setsCount =
        exercise.exercise?.muscleGroups?.filter((muscleGroup) => muscleGroup._id === group._id).length || 0;

      sets += setsCount;

      if (exercise.exercise?.muscleGroups?.some((groupToFilter) => groupToFilter._id === group._id)) {
        repeats += exercise.repeats;
      }
    });

    if (sets) groups.push({ title, sets, repeats });
  });

  return groups.sort((a, b) => b.repeats - a.repeats);
}

export function isPrevExerciseSame(exercises: IExerciseDone[], index: number, id?: string) {
  return id && exercises[index - 1] ? id === exercises[index - 1].exercise?._id : false;
}

export function isSetCreatable(choosenExercises: IExerciseChoosen[], index: number, id?: string) {
  return !isPrevExerciseSame(choosenExercises, index, id) && index > 0 && index + 1 === choosenExercises.length;
}

export function getExercisePassingTitle(index: number, isCurrent: boolean, count: number, exercise: IExerciseDone) {
  return `${index}${isCurrent ? ` из ${count}.` : `.`} ${exercise.exercise?.title || 'Упражнение удалено'}${exercise.weight ? ` ${exercise.weight} кг.` : `.`}`;
}

export function generateTimeline(exercises: IExerciseDone[], start: Date | string | null, ratio: number) {
  const allSteps: ITimelineStep[] = [];

  exercises.forEach((exercise) => {
    allSteps.push({
      left:
        exercise.dateUpdated && exercise.duration && start
          ? (new Date(exercise.dateUpdated).getTime() - exercise.duration * 1000 - new Date(start).getTime()) / ratio
          : 0,
      right:
        exercise.dateUpdated && start
          ? (new Date(exercise.dateUpdated).getTime() - new Date(start).getTime()) / ratio
          : 0,
      type: 'exercise',
    });
  });

  allSteps.forEach((step, index) => {
    allSteps.push({
      left: index === 0 ? 0 : allSteps[index - 1].right,
      right: step.left,
      type: 'rest',
    });
  });

  return allSteps.sort((a, b) => a.left - b.left);
}

export function filterExercisesByTitleAndMuscleGroup(
  exercises: IExercise[],
  title: string,
  muscleGroup: string,
  user?: IUser
) {
  return exercises.filter((exercise) => {
    const isEquipmentMatches = checkIsUserEquipmentMatches(exercise, user);

    const titleFilter = exercise.title.toLowerCase().includes(title.toLocaleLowerCase()) && isEquipmentMatches;
    const muscleGroupFilter = exercise.muscleGroups?.some((group) => group._id === muscleGroup) && isEquipmentMatches;

    return muscleGroup ? muscleGroupFilter && titleFilter : titleFilter;
  });
}

export function getAvailableExerciseWeights(exercise: IExercise, user: IUser) {
  const equipments = user.equipments?.filter((equipment) =>
    exercise.equipmentForWeight?.some((eq) => eq.title === equipment.equipment?.title)
  );

  if (!equipments?.length) return undefined;

  const weights = getWeightsForUserEquipment(equipments);

  return exercise.isWeightsRequired ? weights : [0, ...weights];
}

export function getExercisesToChooseDefaultWeight(exercises: IExercise[], userEquipments?: IUserEquipment[]) {
  const equipmentsForWeight = userEquipments?.filter((eq) => eq.weights?.length);
  const exercisesWithWeight = exercises.filter((ex) => ex.isWeights);

  const tableData = exercisesWithWeight.map((ex) => {
    const exerciseEquipment = ex.equipmentForWeight
      ?.filter((eq) => equipmentsForWeight?.some((equipment) => equipment.equipment?.title === eq.title))
      .map((exToMap) => exToMap.title);

    const availableEqupment = equipmentsForWeight?.filter((eq) =>
      exerciseEquipment?.some((equipmentTitle) => equipmentTitle === eq.equipment?.title)
    );

    const options = getWeightsForUserEquipment(availableEqupment);

    return { _id: ex._id, title: ex.title, options: ex.isWeightsRequired ? options : [0, ...options] };
  });

  return tableData.filter((ex) => ex.options.length > 1);
}

export function getDefaultExerciseWeight(exercise: IExercise, user: IUser, weights?: number[]) {
  let weight = weights?.length && exercise.isWeightsRequired ? weights[0] : 0;

  if (user.defaultWeights && exercise._id && Object.hasOwn(user.defaultWeights, exercise._id)) {
    weight = user.defaultWeights[exercise._id];
  }

  return weight;
}
