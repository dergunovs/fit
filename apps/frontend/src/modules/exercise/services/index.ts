import { ComputedRef, Ref } from 'vue';
import {
  API_EXERCISE,
  API_EXERCISE_ALL,
  API_EXERCISE_CUSTOM,
  TGetExercisesDTO,
  TGetExercisesCustomDTO,
  TGetExercisesQueryDTO,
  TGetExerciseDTO,
  TPostExerciseDTO,
  TPostExerciseDataDTO,
  TUpdateExerciseDTO,
  TUpdateExerciseDataDTO,
  TDeleteExerciseDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const exerciseService = {
  getMany: (page: Ref<number>) => {
    return useQuery({
      queryKey: [API_EXERCISE, page],
      queryFn: async () => {
        const params: TGetExercisesQueryDTO = { page: page.value };

        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE, { params });

        return data;
      },
    });
  },

  getAll: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_ALL],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesDTO>(API_EXERCISE_ALL);

        return data.data;
      },
    });
  },

  getCustom: () => {
    return useQuery({
      queryKey: [API_EXERCISE, API_EXERCISE_CUSTOM],
      queryFn: async () => {
        const { data } = await api.get<TGetExercisesCustomDTO>(API_EXERCISE_CUSTOM);

        return data.data;
      },
    });
  },

  getOne: (options: object, id?: ComputedRef<string | undefined>) => {
    return useQuery({
      queryKey: [API_EXERCISE, id],
      queryFn: async () => {
        if (!id?.value) return null;

        const { data } = await api.get<TGetExerciseDTO>(`${API_EXERCISE}/${id.value}`);

        return data.data;
      },
      ...options,
    });
  },

  create: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TPostExerciseDataDTO) => {
        const { data } = await api.post<TPostExerciseDTO>(API_EXERCISE, formData);

        return data;
      },
      ...options,
    });
  },

  update: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (formData: TUpdateExerciseDataDTO) => {
        const { data } = await api.patch<TUpdateExerciseDTO>(`${API_EXERCISE}/${formData._id}`, formData);

        return data;
      },
      ...options,
    });
  },

  delete: (options: object) => {
    return useMutation({
      mutationKey: [API_EXERCISE],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteExerciseDTO>(`${API_EXERCISE}/${id}`);

        return data;
      },
      ...options,
    });
  },
};
