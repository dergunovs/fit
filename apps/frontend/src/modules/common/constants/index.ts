import IconExercise from '@/common/icons/nav-exercise.svg?component';
import IconActivity from '@/common/icons/nav-activity.svg?component';
import IconEquipment from '@/common/icons/weight.svg?component';
import IconUser from '@/common/icons/nav-user.svg?component';

import { URL_USER } from '@/user/constants';
import { URL_EXERCISE } from '@/exercise/constants';
import { URL_ACTIVITY_ADMIN } from '@/activity/constants';
import { URL_EQUIPMENT } from '@/equipment/constants';
import { INavItem } from '@/common/interface';
import { URL_SETUP } from '@/auth/constants';

export const NAV_ITEMS: INavItem[] = [
  { _id: '1', url: URL_EXERCISE, title: 'Упражнения', icon: IconExercise },
  { _id: '2', url: URL_ACTIVITY_ADMIN, title: 'Занятия', icon: IconActivity },
  { _id: '3', url: URL_EQUIPMENT, title: 'Оборудование', icon: IconEquipment },
  { _id: '4', url: URL_USER, title: 'Пользователи', icon: IconUser },
];

export const UPDATE_BUTTON_TEXT = 'Обновить';
export const CREATE_BUTTON_TEXT = 'Создать';

export const URL_HOME = '/';
export const URL_ERROR = '/404';

export const GUEST_PAGES = [URL_HOME, URL_SETUP];
