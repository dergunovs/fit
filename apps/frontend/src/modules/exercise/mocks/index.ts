import { vi } from 'vitest';
import {
  TDeleteExerciseDTO,
  TGetExercisesDTO,
  TPostExerciseDataDTO,
  TPostExerciseDTO,
  TUpdateExerciseDataDTO,
  TUpdateExerciseDTO,
} from 'fitness-tracker-contracts';

import { exerciseService } from '@/exercise/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { EXERCISE_FIXTURE, EXERCISE_FIXTURE_CUSTOM, EXERCISES_FIXTURE } from '@/exercise/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetExercise = vi.spyOn(exerciseService, 'getOne').mockReturnValue(mockQueryReply(EXERCISE_FIXTURE));

const mockGetExercisesData: TGetExercisesDTO = { data: EXERCISES_FIXTURE, total: EXERCISES_FIXTURE.length };

const spyGetExercises = vi
  .spyOn(exerciseService, 'getMany')
  .mockImplementation(() => mockQueryReply(mockGetExercisesData));

const spyGetExercisesAll = vi
  .spyOn(exerciseService, 'getAll')
  .mockImplementation(() => mockQueryReply(EXERCISES_FIXTURE));

const spyGetExercisesCustom = vi
  .spyOn(exerciseService, 'getCustom')
  .mockImplementation(() => mockQueryReply([EXERCISE_FIXTURE_CUSTOM]));

const spyCreateExercise = vi.fn();
const spyUpdateExercise = vi.fn();
const spyDeleteExercise = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(exerciseService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostExerciseDTO, TPostExerciseDataDTO>(spyCreateExercise);
});

vi.spyOn(exerciseService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateExerciseDTO, TUpdateExerciseDataDTO>(spyUpdateExercise);
});

vi.spyOn(exerciseService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteExerciseDTO, string>(spyDeleteExercise);
});

export {
  spyGetExercise,
  spyGetExercises,
  spyGetExercisesCustom,
  spyGetExercisesAll,
  spyCreateExercise,
  spyUpdateExercise,
  spyDeleteExercise,
  mockOnSuccess,
  mockGetExercisesData,
};
