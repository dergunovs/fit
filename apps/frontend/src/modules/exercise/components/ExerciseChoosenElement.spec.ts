import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import ExerciseChoosenElement from './ExerciseChoosenElement.vue';

import { wrapperFactory } from '@/common/test';
import { EXERCISE_CHOOSEN_FIXTURE } from '@/exercise/fixtures';

const exerciseTitle = dataTest('exercise-title');
const exerciseDelete = dataTest('exercise-delete');
const exerciseCreateSet = dataTest('exercise-create-set');

const index = 1;

let wrapper: VueWrapper<InstanceType<typeof ExerciseChoosenElement>>;

beforeEach(() => {
  wrapper = wrapperFactory(ExerciseChoosenElement, { exercise: EXERCISE_CHOOSEN_FIXTURE, index, isSetCreatable: true });
});

enableAutoUnmount(afterEach);

describe('ExerciseChoosenElement', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ExerciseChoosenElement)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('shows title', async () => {
    expect(wrapper.find(exerciseTitle).text()).toBe(
      `${index}. ${EXERCISE_CHOOSEN_FIXTURE.exercise?.title} x${EXERCISE_CHOOSEN_FIXTURE.repeats}`
    );
  });

  it('shows create set button', async () => {
    expect(wrapper.find(exerciseCreateSet).exists()).toBe(true);

    await wrapper.setProps({ isSetCreatable: false });

    expect(wrapper.find(exerciseCreateSet).exists()).toBe(false);
  });

  it('emits delete choosen exercise', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('delete');

    await wrapper.find(exerciseDelete).trigger('click');

    expect(wrapper.emitted('delete')).toHaveLength(1);
    expect(wrapper.emitted()['delete'][0]).toStrictEqual([EXERCISE_CHOOSEN_FIXTURE._id]);
  });

  it('emits create set', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('createSet');

    await wrapper.find(exerciseCreateSet).trigger('click');

    expect(wrapper.emitted('createSet')).toHaveLength(1);
  });
});
