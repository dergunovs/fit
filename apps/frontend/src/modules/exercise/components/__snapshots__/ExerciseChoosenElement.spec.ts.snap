// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExerciseChoosenElement > matches snapshot 1`] = `
"<div class="_exercise_133bad"><span data-test="exercise-title">1. Отжимание от скамьи x12</span>
  <div><button layout="plain" data-test="exercise-create-set"> +Сет </button>
    <ui-close-stub issmall="true" isdelete="false" data-test="exercise-delete"></ui-close-stub>
  </div>
</div>"
`;
